File: backend/bin/main/infrastructure/external/AiClient.kt
/*
 * ॐ नमः शिवाय
 * By: Sheersh Bhatnagar
 * Date: 24/04/26 18:19
 */

package dev.sheershbhatnagar.ai_assistant.infrastructure.external

import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.json.*

class AiClient(private val client: HttpClient) {

    suspend fun callGemini(modelName: String, apiKey: String, prompt: String): String {
        val url = "https://generativelanguage.googleapis.com/v1beta/models/$modelName:generateContent"

        val response = client.post(url) {
            header("x-goog-api-key", apiKey)
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject {
                putJsonArray("contents") {
                    addJsonObject {
                        putJsonArray("parts") {
                            addJsonObject { put("text", prompt) }
                        }
                    }
                }
            })
        }

        // Basic parsing of Gemini's specific JSON structure
        val jsonResponse = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        return jsonResponse["candidates"]?.jsonArray?.get(0)?.jsonObject?.get("content")
            ?.jsonObject?.get("parts")?.jsonArray?.get(0)?.jsonObject?.get("text")
            ?.jsonPrimitive?.content ?: "No response from Gemini"
    }

    suspend fun callOpenAI(modelName: String, apiKey: String, prompt: String): String {
        val response = client.post("https://api.openai.com/v1/chat/completions") {
            header(HttpHeaders.Authorization, "Bearer $apiKey")
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject {
                put("model", modelName)
                putJsonArray("messages") {
                    addJsonObject {
                        put("role", "user")
                        put("content", prompt)
                    }
                }
            })
        }

        val jsonResponse = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        return jsonResponse["choices"]?.jsonArray?.get(0)?.jsonObject?.get("message")
            ?.jsonObject?.get("content")?.jsonPrimitive?.content ?: "No response from OpenAI"
    }
}

File: backend/app/utils/ai_client.py
import httpx

class AiClient:
    async def call_gemini(self, model_name: str, api_key: str, prompt: str) -> str:
        url = f"https://generativelanguage.googleapis.com/v1beta/models/{model_name}:generateContent"
        headers = {
            "x-goog-api-key": api_key,
            "Content-Type": "application/json"
        }
        body = {
            "contents": [
                {
                    "parts": [
                        {"text": prompt}
                    ]
                }
            ]
        }
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(url, headers=headers, json=body, timeout=30.0)
                response.raise_for_status()
                data = response.json()
                # candidates[0].content.parts[0].text
                return data["candidates"][0]["content"]["parts"][0]["text"]
            except Exception as e:
                # Log or rethrow
                print(f"Error calling Gemini: {e}")
                if "response" in locals() and response is not None:
                    print(f"Response content: {response.text}")
                return f"Error calling Gemini: {str(e)}"

    async def call_openai(self, model_name: str, api_key: str, prompt: str) -> str:
        url = "https://api.openai.com/v1/chat/completions"
        headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }
        body = {
            "model": model_name,
            "messages": [
                {
                    "role": "user",
                    "content": prompt
                }
            ]
        }
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(url, headers=headers, json=body, timeout=30.0)
                response.raise_for_status()
                data = response.json()
                # choices[0].message.content
                return data["choices"][0]["message"]["content"]
            except Exception as e:
                # Log or rethrow
                print(f"Error calling OpenAI: {e}")
                if "response" in locals() and response is not None:
                    print(f"Response content: {response.text}")
                return f"Error calling OpenAI: {str(e)}"
